memory-db/src/lib.rs

```diff
@@ -160,19 +160,151 @@ impl<H, KF, T> Eq for MemoryDB<H, KF, T>
 pub trait KeyFunction<H: KeyHasher> {
 	type Key: Send + Sync + Clone + hash::Hash + Eq;
 
-	fn key(hash: &H::Out, prefix: &[u8]) -> Self::Key;
+	fn key(hash: &H::Out, prefix: Prefix) -> Self::Key;
+}
+
+
+/// Make database key from hash and prefix.
+///
+/// Prefix could be ordered, but key hash appending will break this property.
+/// If true ordering is needed we would need a fix length encoding for prefix
+/// (so a maximum length and additional byte (or chunk of prefix encoded)).
+pub fn prefixed_key<H: KeyHasher>(key: &H::Out, prefix: Prefix) -> Vec<u8> {
+	let mut prefixed_key = Vec::with_capacity(key.as_ref().len() + prefix.0.len() + 1);
+	prefixed_key.extend_from_slice(prefix.0);
+	if (prefix.1).0 > 0 {
+		prefixed_key.push((prefix.1).1);
+		prefixed_key.push((prefix.1).0); // put size to avoid any possible collision
+	}
+	prefixed_key.extend_from_slice(key.as_ref());
+	prefixed_key
+}
+
+#[test]
+fn test_ordered_prefixed_key() {
+	use keccak_hasher::KeccakHasher;
+	#[derive(Debug, Clone, Default)]
+	struct Opt;
+	impl IntoIterator for Opt {
+		type Item = usize;
+		type IntoIter = std::iter::Repeat<usize>;
+		fn into_iter(self) -> Self::IntoIter {
+			std::iter::repeat(1)
+		}
+	}
+	impl OrderedPrefixedKeyChunk for Opt { }
+	let prefixes = vec![
+		(vec![],(0u8, 0x00u8), [1u8;32]),
+		(vec![],(1u8, 0x00u8), [1u8;32]),
+		(vec![0u8],(0,0), [0;32]),
+		(vec![],(1u8, 0x10u8), [1u8;32]),
+		(vec![0x10u8],(0,0), [0;32]),
+		(vec![0x11u8],(0,0), [0;32]),
+	];
+	let oc: std::collections::BTreeSet<_> = prefixes.iter().map(|(p1, p2, k)| {
+		ordered_prefixed_key::<KeccakHasher, Opt>(&k.clone().into(), (&p1[..], p2.clone()))
+	}).collect();
+
+	for ((p1, p2, k),b) in prefixes.iter().zip(oc.iter()) {
+		assert_eq!(
+			&ordered_prefixed_key::<KeccakHasher, Opt>(&k.clone().into(), (&p1[..], p2.clone())),
+			b,
+		);
+	}
+}
+/// Parameterize the size of successive chunks when using
+/// `ordered_prefixed_key`.
+/// Length of each chunk is minimum 1, maximum 254.
+pub trait OrderedPrefixedKeyChunk: IntoIterator<Item = usize> + Clone + MaybeDebug + Default {
```

Comment:
Whats a "chunk" and how is it formed? Furthermore, is this trait really needed? Does not seem to be implemented anywhere else.